Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/AbiSpecific.kt

```diff
@@ -0,0 +1,132 @@
+package org.jetbrains.kotlin.native.interop.gen
+
+import org.jetbrains.kotlin.konan.target.KonanTarget
+import org.jetbrains.kotlin.native.interop.indexer.*
+
+/**
+ * objc_msgSend*_stret functions must be used when return value is returned through memory
+ * pointed by implicit argument, which is passed on the register that would otherwise be used for receiver.
+ *
+ * The entire implementation is just the real ABI approximation which is enough for practical cases.
+ */
+internal fun Type.isStret(target: KonanTarget): Boolean {
+    val unwrappedType = this.unwrapTypedefs()
+    val abiInfo: ObjCAbiInfo = when (target) {
+        KonanTarget.IOS_ARM64,
+        KonanTarget.TVOS_ARM64 -> DarwinArm64AbiInfo()
+
+        KonanTarget.IOS_X64,
+        KonanTarget.MACOS_X64,
+        KonanTarget.WATCHOS_X64,
+        KonanTarget.TVOS_X64 -> DarwinX64AbiInfo()
+
+        KonanTarget.WATCHOS_X86 -> DarwinX86AbiInfo()
+
+        KonanTarget.IOS_ARM32,
+        KonanTarget.WATCHOS_ARM32 -> DarwinArm32AbiInfo(target)
+
+        else -> error("Cannot generate ObjC stubs for $target.")
+    }
+    return abiInfo.shouldUseStret(unwrappedType)
+}
+
+/**
+ * Provides ABI-specific information about target for Objective C interop.
+ */
+interface ObjCAbiInfo {
+    fun shouldUseStret(returnType: Type): Boolean
+}
+
+class DarwinX64AbiInfo : ObjCAbiInfo {
+    override fun shouldUseStret(returnType: Type): Boolean {
+        return when (returnType) {
+            is RecordType -> returnType.decl.def!!.size > 16 || returnType.hasUnalignedMembers()
+            else -> false
+        }
+    }
+}
+
+class DarwinX86AbiInfo : ObjCAbiInfo {
+    override fun shouldUseStret(returnType: Type): Boolean {
+        // See: https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/LowLevelABI/130-IA-32_Function_Calling_Conventions/IA32.html#//apple_ref/doc/uid/TP40002492-SW5
+        // Structures 1 or 2 bytes in size are placed in EAX.
+        // Structures 4 or 8 bytes in size are placed in: EAX and EDX.
```

Comment:
Not sure if the comment is correct. Isn't `struct S { float x; }` returned on FPU stack?

Reply:
You are correct.
Turns out, this kind of structs is a special case. https://github.com/llvm/llvm-project/blob/6c8a34ed9b49704bdd60838143047c62ba9f2502/clang/lib/CodeGen/TargetInfo.cpp#L1438